### PR TITLE
EVG-15254: Add first 10 failing tests to test results DB doc

### PR DESCRIPTION
**Pull request by julianedwards**
JIRA: https://jira.mongodb.org/browse/EVG-15254

This is to optimize fetching of a sample of the failing tests for a given task, used by the UI for hover pop-ups.

### CODE REVIEW

File: model/test_results.go [hunk 1]
```diff
@@ -195,6 +208,44 @@ func (t *TestResults) Append(ctx context.Context, results []TestResult) error {
...
+		if strings.Contains(strings.ToLower(results[i].Status), "fail") {
+			t.FailedTestsSample = append(t.FailedTestsSample, results[i].getDisplayName())
+			update = true
+		}
+	}
+	if !update {
+		return nil
+	}
+
+	if t.ID == "" {
```
Comment (Kimchelly) [inline model/test_results.go:223]: In what case would the ID not be set? It seems like the ID is generated with `CreateTestResultsRecord` and calling `Append from the gRPC route populates the ID when retrieving it from the DB.

File: model/test_results.go [hunk 2]
```diff
@@ -195,6 +208,44 @@ func (t *TestResults) Append(ctx context.Context, results []TestResult) error {
 	return errors.Wrap(bucket.Put(ctx, testResultsCollection, bytes.NewReader(data)), "uploading test results")
 }
 
+func (t *TestResults) appendToFailedTestsSample(ctx context.Context, results []TestResult) error {
+	var update bool
+	for i := 0; i < len(results) && len(t.FailedTestsSample) < failedTestsSampleSize; i++ {
+		if strings.Contains(strings.ToLower(results[i].Status), "fail") {
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/test_results.go:214]: I wanted to leave it flexible, since cedar technically does not know about Evg's internals.

**Comment by julianedwards** [comment]
> I was wondering, in the interest of minimizing the amount of stuff in the db:
> 
> 1. Is there any way to only store for the latest execution of a task?
> 2. is there any way to store only 10 results for a display task?
> 
> I mean, it's not the biggest deal to add a short string array for every test result, but I wanted to raise the question since it seems to me the UI needs would still be met.

1. It would require more work to store only for the latest execution (we would have to retroactively go back and delete the failing test names from the previous executions).

2. I believe the UI needs failing tests at the execution task level as well as the display task level. Also, this wouldn't be possible since we don't actually store display tasks (rather it is an optional field that is part of each execution task).

File: model/test_results_test.go [hunk 3]
```diff
@@ -263,14 +265,36 @@ func TestTestResultsAppend(t *testing.T) {
...
 		assert.Equal(t, results, savedResults.Results)
+		var saved TestResults
+		require.NoError(t, db.Collection(testResultsCollection).FindOne(ctx, bson.M{"_id": tr.ID}).Decode(&saved))
+		assert.Empty(t, tr.FailedTestsSample)
+
+		failedResults := make([]TestResult, 20)
+		for i := 0; i < 10; i++ {
+			failedResults[i] = getTestResult()
+			failedResults[i].Status = "fail"
+		}
```
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/test_results_test.go:281]: Fixed it.

**Review by Kimchelly — APPROVED** [review]

**Pull request merged**